_source/_posts/2022-07-21-jhipster-ci-cd.md

```diff
@@ -0,0 +1,865 @@
+- Organization: **your-dockerhub-username**
+- Image: **your-dockerhub-username/store**
+- Artifact Constraints: Choose **Define new artifact**
+
+In the _Expected Artifact_ form, set the following values:
+
+- Display Name: **store-image**
+- Account: **docker-registry**
+- Docker image: Fully qualified **index.docker.io/your-dockerhub-username/store**
+- Use prior execution: **yes**
+- Use default artifact: **yes** (set the same account and object path as before)
+
+Click **Save Artifact**.
+
+{% img blog/jhipster-ci-cd/sp-trigger-artifact.png alt:"Spinnaker pipeline trigger" width:"700" %}{: .center-image }
+
+**IMPORTANT NOTE**: Spinnaker provides a mechanism to [override](https://spinnaker.io/docs/guides/user/kubernetes-v2/deploy-manifest/#override-artifacts) the version of Docker images, Kubernetes ConfigMaps, and Secrets in manifests, injecting new versions when one of these objects exist in the pipeline context. However, Spinnaker documentation does not have a clear example of the trigger configuration for the docker image to be available in the pipeline context. To make it happen, you must _set it as an artifact constraint in the trigger_. If the artifact constraint is not defined, the original image version in the manifests will be deployed, instead of the new image that triggered the pipeline. If no version is in the manifests, `latest` will be the default.
+
+Add a second Docker registry trigger for the gateway image pushes and click **Save Changes**.  
+
+### Configure the deployment stages
+
+The store pipeline will execute the following stages:
+
+- Deploy the application ConfigMap
+- Deploy the JHipster Registry
+- Deploy the gateway database PostgreSQL
+- Deploy the store clustered database MongoDB
+- Deploy the gateway microservice
+- Deploy the store microservice
+- Enable the gateway Service
+- Enable the store Service
+
+The Kubernetes configuration created in the folder `jhipster-ci-cd/kubernetes` must be pushed to a GitHub repository, so Spinnaker can access it using the GitHub artifact account created before. But before the push, Edit `kubernetes/store-k8s/store-deployment.yml` and set the image name with the fully qualified address:
+
+```yml
+- name: store-app
+  image: index.docker.io/indiepopart/store
+```
+Do the same for `kubernetes/gateway-k8s/gateway-deployment.yml`. This is also required for the artifact substitution to work.
+
+Sign in to GitHub and create a public repository `jhipster-k8s`. Follow the instructions to push the existing `kubernetes` folder from your local using the command line.
+
+**IMPORTANT NOTE**: You will be pushing plain secrets contained in the application and Kubernetes configuration. To avoid this, you can run the JHipster registry locally to encrypt application configuration, and also set up `kubeseal` for Kubernetes secrets encryption. The process is described in a previous post [Kubernetes to the Cloud with Spring Boot and JHipster](/blog/2021/06/01/kubernetes-spring-boot-jhipster).
```

Comment:
```suggestion
**IMPORTANT NOTE**: At this point, you will be pushing plain secrets contained in the application and Kubernetes configuration. To avoid this insecure anti-pattern, you can run the JHipster registry locally to encrypt application configuration, and also set up `kubeseal` for Kubernetes secrets encryption. The process is described in a previous post [Kubernetes to the Cloud with Spring Boot and JHipster](/blog/2021/06/01/kubernetes-spring-boot-jhipster).
```